boto3/dynamodb/transform.py

```diff
@@ -0,0 +1,188 @@
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from collections import Mapping
+import functools
+
+from boto3.dynamodb.types import TypeSerializer, TypeDeserializer
+from boto3.dynamodb.conditions import ConditionBase
+from boto3.dynamodb.conditions import ConditionExpressionBuilder
+
+
+def register_high_level_interface(base_classes, **kwargs):
+    # This assumes the base class is the base resource object
+    resource_base_class = base_classes[0]
+
+    class DynamoDBHighLevelResource(resource_base_class):
+        def __init__(self, *args, **kwargs):
+            super(DynamoDBHighLevelResource, self).__init__(*args, **kwargs)
+            # Apply the handler that generates condition expressions including
+            # placeholders.
+
+            self.meta.client.meta.events.register(
+                'before-parameter-build.dynamodb',
+                transform_condition_expressions,
+                unique_id='dynamodb-condition-expression')
+
+            # Apply the handler that serializes the request from python
+            # types to dynamodb types.
+            self.meta.client.meta.events.register(
+                'before-parameter-build.dynamodb',
+                transform_attribute_value_input,
+                unique_id='dynamodb-attr-value-input')
+
+            # Apply the handler that deserializes the response from dynamodb
+            # types to python types.
+            self.meta.client.meta.events.register(
+                'after-call.dynamodb', transform_attribute_value_output,
+                unique_id='dynamodb-attr-value-output')
+
+    base_classes[0] = DynamoDBHighLevelResource
+
+
+def transform_condition_expressions(params, model, **kwargs):
+    transformer = ParameterTransformer()
+    condition_builder = ConditionExpressionBuilder()
+
+    def inject_condition_expression(value, transformer, condition_builder,
+                                    original_params, model, placeholder_names,
+                                    placeholder_values,
+                                    is_key_condition=False):
+        if isinstance(value, ConditionBase):
+            # Create a conditional expression string with placeholders
+            # for the provided condition.
+            expr, attr_names, attr_values = condition_builder.build_expression(
+                value, is_key_condition=is_key_condition)
+
+            placeholder_names.update(attr_names)
+            placeholder_values.update(attr_values)
+
+            return expr
+        # Use the user provided value if it is not a ConditonBase object.
+        return value
+
+    generated_names = {}
+    generated_values = {}
+
+    # Apply transformation for conditon and key conditon expression parameters.
+    cond_exp_transformation = functools.partial(
```

Comment:
What do you think about creating a class that encapsulates the inject_condition_expression and avoids the partial?  I think that would help simplify the code and is something we could unit test directly, i.e

```
class SomethingSomethingInjector(object):
   def __init__(self, transformer, condition_builder, ...):
       ...
    def __call__(self, value):
       ... transform the value ....
```


It would also help because we could then document the class.